File: algorithms/11_golang/data_structures/sparse_table.go
package main

import "fmt"

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

type SparseTable struct {
	st  [][]int
	log []int
}

func NewSparseTable(arr []int) *SparseTable {
	n := len(arr)
	log := make([]int, n+1)
	for i := 2; i <= n; i++ {
		log[i] = log[i/2] + 1
	}
	k := log[n] + 1
	st := make([][]int, n)
	for i := 0; i < n; i++ {
		st[i] = make([]int, k)
		st[i][0] = arr[i]
	}
	for j := 1; j < k; j++ {
		for i := 0; i+(1<<j) <= n; i++ {
			st[i][j] = minInt(st[i][j-1], st[i+(1<<(j-1))][j-1])
		}
	}
	return &SparseTable{st: st, log: log}
}

func (s *SparseTable) Query(l, r int) int {
	j := s.log[r-l+1]
	return minInt(s.st[l][j], s.st[r-(1<<j)+1][j])
}

func main() {
	arr := []int{2, 5, 1, 4, 9, 3}
	st := NewSparseTable(arr)
	if st.Query(0, 5) != 1 || st.Query(1, 3) != 1 || st.Query(4, 5) != 3 || st.Query(0, 0) != 2 {
		panic("SparseTable: RMQ mismatch")
	}
	fmt.Println("[Go SparseTable] Range minimum query verified.")
}

File: algorithms/06_javascript/data_structures/sparse_table.js
class SparseTable {
  constructor(arr) {
    this.n = arr.length;
    this.log = new Array(this.n + 1).fill(0);
    for (let i = 2; i <= this.n; i++) this.log[i] = this.log[i >> 1] + 1;
    const k = this.log[this.n] + 1;
    this.st = Array.from({ length: this.n }, () => new Array(k).fill(0));
    for (let i = 0; i < this.n; i++) this.st[i][0] = arr[i];
    for (let j = 1; j < k; j++)
      for (let i = 0; i + (1 << j) <= this.n; i++)
        this.st[i][j] = Math.min(this.st[i][j - 1], this.st[i + (1 << (j - 1))][j - 1]);
  }
  query(l, r) {
    const j = this.log[r - l + 1];
    return Math.min(this.st[l][j], this.st[r - (1 << j) + 1][j]);
  }
}

module.exports = { SparseTable };

if (require.main === module) {
  const st = new SparseTable([4, 2, 6, 1, 9, 3]);
  if (st.query(1, 4) !== 1) throw new Error("RMQ query failed");
  if (st.query(0, 2) !== 2) throw new Error("RMQ query failed");
  console.log("[JavaScript Sparse Table] Range minimum query verified");
}
